flang/lib/Optimizer/OpenMP/GenericLoopConversion.cpp

```diff
@@ -343,6 +343,115 @@ class GenericLoopConversionPattern
   }
 };
 
+/// According to the spec (v5.2, p340, 36):
```

Comment:
The PR needs more testing, working on that. I opened at this point for early feedback in case there are major issues with the approach taken.